### DISCUSSION TITLE
adroino uno wsb 24bit help

### DISCUSSION THREAD

**Discussion by bujo-hub** [community/community#134565] [Programming Help]
### Body

Hello, I'm new here and would like to start a project. Unfortunately, I don't know much about it 
I would like to write a code for my ardoino uno 
for the wsb 24bit led 
that should have the function of 
step 1 
DRL start scan white 
to 40% from left to right then back then again from left to right but with 90% luminosity
Then an effect should come slowly with 3 pulses that turn blue 
and after 3 pulses it should glow permanently blue without any effect on 90% luminosity

Can someone send me such a code

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ebndev** [comment]
Hi @bujo-hub, thanks for being a part of the GitHub Community!

It looks like you'll need a lot of help with starting your Arduino project. It's possible that another GitHub user might have run into this same issue and can help, but the [Arduino forum](https://forum.arduino.cc/?_gl=1*lfrla3*_ga*MTg2Mzk0MDQ1Ny4xNzE3MDkzMTA2*_ga_NEXN8H46L5*MTcyMjk4MTQwMi4yLjAuMTcyMjk4MTQxMC4wLjAuMTgzNDY0Nzk0MA..*_fplc*ZlFmTm1LTUNyOVBCc2pYYnE2clZTelNlcEclMkZadHJMNHNXWmJyNCUyRnk3OVJjV3NXJTJGOWlydnE5cXJ5dExQNWxlR2ZzbSUyRlZmR20xV2dHeGZ2Ylh3eXVPQnJTT2M5T3hoQnRqJTJGY0MyTVVXdXBnV3ZLN1p3MiUyQjNxYjFsJTJGdGl0YkElM0QlM0Q.*_gcl_au*MjAxMzI1MDMxMi4xNzE3MDkzMTM5*FPAU*MjAxMzI1MDMxMi4xNzE3MDkzMTM5) may be a better place to get started. They have tons of example projects and code you can browse as well!

We want to make sure you’re getting the best support you can, but this space may not be the right place for this particular topic.

Best of luck!

**Comment by geniuszly** [comment]
```cpp
#include <Adafruit_NeoPixel.h>

#define PIN 6 // Pin where the LED strip is connected
#define NUM_LEDS 30 // Number of LEDs in the strip

Adafruit_NeoPixel strip = Adafruit_NeoPixel(NUM_LEDS, PIN, NEO_GRB + NEO_KHZ800);

void setup() {
  strip.begin();
  strip.show(); // Initialize all pixels to 'off'
}

void loop() {
  DRLScan(40);  // Step 1: DRL start scan at 40% brightness
  delay(500);
  DRLScan(90);  // Step 2: DRL scan at 90% brightness
  delay(500);
  BluePulse();  // Step 3: Blue pulses
  PermanentBlue(90); // Step 4: Permanent blue at 90% brightness
}

void showAuthor() {
  int purpleColor = strip.Color(128, 0, 128); // Purple color
  strip.fill(purpleColor);
  strip.show();
  delay(3000); // Purple color display for 3 seconds
  strip.clear(); // Cleaning after display
  strip.show();
}

void DRLScan(uint8_t brightness) {
  int whiteColor = strip.Color((brightness * 255) / 100, (brightness * 255) / 100, (brightness * 255) / 100);
  
  // Scan from left to right
  for(int i = 0; i < NUM_LEDS; i++) {
    strip.setPixelColor(i, whiteColor);
    strip.show();
    delay(50); // Adjust delay for speed
  }
  
  // Scan from right to left
  for(int i = NUM_LEDS - 1; i >= 0; i--) {
    strip.setPixelColor(i, whiteColor);
    strip.show();
    delay(50); // Adjust delay for speed
  }
}

void BluePulse() {
  int blueColor = strip.Color(0, 0, 255);

  for(int i = 0; i < 3; i++) {
    for(int j = 0; j <= 255; j += 5) {
      strip.fill(strip.Color(0, 0, j));
      strip.show();
      delay(10); // Adjust delay for speed
    }
    for(int j = 255; j >= 0; j -= 5) {
      strip.fill(strip.Color(0, 0, j));
      strip.show();
      delay(10); // Adjust delay for speed
    }
  }
}

void PermanentBlue(uint8_t brightness) {
  int blueColor = strip.Color(0, 0, (brightness * 255) / 100);
  strip.fill(blueColor);
  strip.show();
}
```
`strip` - An object is being created to control the LEDs for class Adafruit_NeoPixel
`setup` - initialization function
`loop` - function of the main program cycle
`showAuthor` - function for displaying the author's signature (lights up in purple for 3 seconds after switching on)
`DRLScan` - Function for DRL scanning
`BluePulse` - Function for creating blue pulses
`PermanentBlue` - Function for constant blue glow

**Reply by bujo-hub** [reply]
thank you for your writing

I've been able to build a version so far, but I can't get any further into a required field

This effect is skipped, I don't know why

This effect should appear after the white scann

void setup() {
  Serial.begin(9600); // Initialize serial communication  
  strip.begin(); // Initialize the NeoPixel strip  
  strip.setBrightness(102); // Set brightness to 40%
  strip.show(); // Turn off all LEDs initially
  
  sparkEffect(3000, 20);  // Run spark effect for 3 seconds with 20% chance of sparkling  
  whiteScanner(1);        // Run white scanner effect for 1 round  
  turnOffInSteps();       // Turn off LEDs in steps of four  
  whitePulse(2);          // Run white pulse effect for 2 rounds  
  blueFadeIn();           // Blue fade-in effect  
  setAllBlue();           // Set all LEDs to blue permanently  

-------------------------------------------------------------------------------
// Function to turn off LEDs in steps of four from left to right  
void turnOffInSteps() {
  for (int i = 0; i < NUM_LEDS; i += 4) {
    for (int j = i; j < i + 4 && j < NUM_LEDS; j++) {
      strip.setPixelColor(j, strip.Color(0, 0, 0)); // Turn off current LED
    }
    strip.show(); // Update the strip to reflect changes  
    delay(200); // Delay between turning off each group of four LEDs
    
    // Optional: Print the current state for debugging  
    Serial.print("Turning off LEDs from ");
    Serial.print(i);
    Serial.print(" to ");
    Serial.println(i + 4);
  }


----------------------------------------------------------------------------------------------


#include <Adafruit_NeoPixel.h>

// Define the number of LEDs and the data pin
#define NUM_LEDS 24 // Corrected to 24 LEDs
#define DATA_PIN 6

// Create an instance of the Adafruit_NeoPixel class
Adafruit_NeoPixel strip = Adafruit_NeoPixel(NUM_LEDS, DATA_PIN, NEO_GRB + NEO_KHZ800);

void setup() {
  strip.begin(); // Initialize the NeoPixel strip
  strip.setBrightness(102); // Set brightness to 40%
  strip.show(); // Turn off all LEDs initially
  
  sparkEffect(3000, 20);  // Run spark effect for 3 seconds with 20% chance of sparkling
  whiteScanner(1);        // Run white scanner effect for 1 round
  whitePulse(2);          // Run white pulse effect for 2 rounds
  blueFadeIn();           // Blue fade-in effect
  setAllBlue();           // Set all LEDs to blue permanently
  
  // Optionally, call the turnOffInSteps here if you want to turn off the LEDs after setting them to blue
  // turnOffInSteps();  // Turn off LEDs in steps of four (optional)
}

void loop() {
  // If no repetition is needed, keep this empty or remove it.
  // You could also add a small delay here if you want a pause before the next cycle.
}

// Function for spark effect
void sparkEffect(int duration, int chanceOfSparkle) {
  int startTime = millis();
  while (millis() - startTime < duration) {
    strip.clear(); // Turn off all LEDs before updating
    
    // Randomly decide which LEDs to sparkle
    for (int i = 0; i < NUM_LEDS; i++) {
      if (random(100) < chanceOfSparkle) {
        strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set random LED to full white for sparkle
      }
    }

    strip.show();
    delay(50); // Delay between updates
  }
  strip.clear(); // Ensure all LEDs are off after the effect
  strip.show();
}

// Function for white droplet effect
void whiteDroplet() {
  for (int i = 0; i < NUM_LEDS; i++) {
    strip.clear(); // Turn off all LEDs
    strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
    strip.show();
    delay(50);
    strip.setPixelColor(i, strip.Color(153, 153, 153)); // Set it to 60% white
    strip.show();
    delay(50);
    strip.setPixelColor(i, strip.Color(102, 102, 102)); // Set it to 40% white
    strip.show();
  }
}

// Function for the white scanner effect
void whiteScanner(int rounds) {
  for (int r = 0; r < rounds; r++) {
    // Forward scan (LEDs won't stay active)
    for (int i = 0; i < NUM_LEDS; i++) {
      strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
      strip.show();
      delay(50);
      if (r == 0) {
        strip.setPixelColor(i, strip.Color(0, 0, 0)); // Set current LED to 40% white
      }
    }
    strip.show();
    delay(80);

    // Backward scan
    for (int i = NUM_LEDS - 1; i >= 0; i--) {
      strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
      strip.show();
      delay(50);
      if (r == 0) {
        strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to 40% white
      }
    }
    strip.show();
    delay(200);
  }
}

// Function for white pulse effect
void whitePulse(int rounds) {
  for (int r = 0; r < rounds; r++) {
    // Brighten up
    for (int brightness = 0; brightness <= 255; brightness += 5) {
      for (int i = 0; i < NUM_LEDS; i++) {
        strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs brighten up
      }
      strip.show();
      delay(10); // Speed of brightening up
    }
    
    // Dim down
    for (int brightness = 255; brightness >= 0; brightness -= 5) {
      for (int i = 0; i < NUM_LEDS; i++) {
        strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs dim down
      }
      strip.show();
      delay(10); // Speed of dimming down
    }
  }
}

// Function for blue fade-in effect
void blueFadeIn() {
  for (int brightness = 0; brightness <= 255; brightness++) {
    for (int i = 0; i < NUM_LEDS; i++) {
      strip.setPixelColor(i, strip.Color(0, 0, brightness)); // Fade in blue
    }
    strip.show();
    delay(40); // Smooth fade-in effect
  }
}

// Function to set all LEDs to blue permanently
void setAllBlue() {
  for (int i = 0; i < NUM_LEDS; i++) {
    strip.setPixelColor(i, strip.Color(0, 0, 255)); // Set LED to full blue
  }
  strip.show();
}

**Reply by geniuszly** [reply]
You have defined `turnOffInSteps` inside the `setup()` function, which is not valid in Arduino/C++. Ensure that each function is defined separately outside of any other functions. 
The `turnOffInSteps` function is missing a closing `}` at the end. 
In the `whiteScanner` function, you set the LEDs back to full white instead of 40% white when scanning backward. Adjust the code according to what you intend. 
You have two `setup()` functions, which isn't allowed. Merge the necessary parts into a single `setup()`.
In `whiteScanner`, you're resetting LEDs to 40% white only in the forward scan, not in the backward scan. If you want them to remain at 40% after the first round, modify the condition accordingly.
My advice to you: improve code readability by ensuring consistent indentation and spacing. This makes it easier to debug and understand.

Here is an example of the correct code, based on how I understood it:
```cpp
#include <Adafruit_NeoPixel.h>

// Define the number of LEDs and the data pin
#define NUM_LEDS 24
#define DATA_PIN 6

// Create an instance of the Adafruit_NeoPixel class
Adafruit_NeoPixel strip = Adafruit_NeoPixel(NUM_LEDS, DATA_PIN, NEO_GRB + NEO_KHZ800);

void setup() {
    Serial.begin(9600); // Initialize serial communication
    strip.begin(); // Initialize the NeoPixel strip
    strip.setBrightness(102); // Set brightness to 40%
    strip.show(); // Turn off all LEDs initially

    sparkEffect(3000, 20); // Run spark effect for 3 seconds with 20% chance of sparkling
    whiteScanner(1); // Run white scanner effect for 1 round
    whitePulse(2); // Run white pulse effect for 2 rounds
    blueFadeIn(); // Blue fade-in effect
    setAllBlue(); // Set all LEDs to blue permanently
}

void loop() {
    // No code needed in loop if effects run only once
}

void sparkEffect(int duration, int chanceOfSparkle) {
    int startTime = millis();
    while (millis() - startTime < duration) {
        strip.clear(); // Turn off all LEDs before updating

        // Randomly decide which LEDs to sparkle
        for (int i = 0; i < NUM_LEDS; i++) {
            if (random(100) < chanceOfSparkle) {
                strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set random LED to full white for sparkle
            }
        }
        strip.show();
        delay(50); // Delay between updates
    }
    strip.clear(); // Ensure all LEDs are off after the effect
    strip.show();
}

void whiteScanner(int rounds) {
    for (int r = 0; r < rounds; r++) {
        // Forward scan
        for (int i = 0; i < NUM_LEDS; i++) {
            strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
            strip.show();
            delay(50);
            if (r == 0) {
                strip.setPixelColor(i, strip.Color(102, 102, 102)); // Set current LED to 40% white
            }
        }
        strip.show();
        delay(80);

        // Backward scan
        for (int i = NUM_LEDS - 1; i >= 0; i--) {
            strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
            strip.show();
            delay(50);
            if (r == 0) {
                strip.setPixelColor(i, strip.Color(102, 102, 102)); // Set current LED to 40% white
            }
        }
        strip.show();
        delay(200);
    }
}

void whitePulse(int rounds) {
    for (int r = 0; r < rounds; r++) {
        // Brighten up
        for (int brightness = 0; brightness <= 255; brightness += 5) {
            for (int i = 0; i < NUM_LEDS; i++) {
                strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs brighten up
            }
            strip.show();
            delay(10); // Speed of brightening up
        }

        // Dim down
        for (int brightness = 255; brightness >= 0; brightness -= 5) {
            for (int i = 0; i < NUM_LEDS; i++) {
                strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs dim down
            }
            strip.show();
            delay(10); // Speed of dimming down
        }
    }
}

void blueFadeIn() {
    for (int brightness = 0; brightness <= 255; brightness++) {
        for (int i = 0; i < NUM_LEDS; i++) {
            strip.setPixelColor(i, strip.Color(0, 0, brightness)); // Fade in blue
        }
        strip.show();
        delay(40); // Smooth fade-in effect
    }
}

void setAllBlue() {
    for (int i = 0; i < NUM_LEDS; i++) {
        strip.setPixelColor(i, strip.Color(0, 0, 255)); // Set LED to full blue
    }
    strip.show();
}

// Function to turn off LEDs in steps of four from left to right
void turnOffInSteps() {
    for (int i = 0; i < NUM_LEDS; i += 4) {
        for (int j = i; j < i + 4 && j < NUM_LEDS; j++) {
            strip.setPixelColor(j, strip.Color(0, 0, 0)); // Turn off current LED
        }
        strip.show(); // Update the strip to reflect changes
        delay(200); // Delay between turning off each group of four LEDs

        // Optional: Print the current state for debugging  
        Serial.print("Turning off LEDs from ");
        Serial.print(i);
        Serial.print(" to ");
        Serial.println(i + 4);
    }
}
```
Corrected LED resetting logic to ensure proper state handling.
This should resolve the issue of the skipped effect and improve the overall functionality of your code.

**Reply by bujo-hub** [reply]
finally I have the code


#include <Adafruit_NeoPixel.h>

// Define the number of LEDs and the data pin
#define NUM_LEDS 24
#define DATA_PIN 6

// Create an instance of the Adafruit_NeoPixel class
Adafruit_NeoPixel strip = Adafruit_NeoPixel(NUM_LEDS, DATA_PIN, NEO_GRB + NEO_KHZ800);

// Forward declarations of functions
void sparkEffect(int duration, int chanceOfSparkle);
void whiteScanner(int rounds);
void whitePulse(int rounds);
void blueFadeIn();
void setAllBlue();

void setup() {
  strip.begin(); // Initialize the NeoPixel strip
  strip.setBrightness(102); // Set brightness to 40%
  strip.show(); // Turn off all LEDs initially
  
  sparkEffect(3000, 20);  // Run spark effect for 3 seconds with 20% chance of sparkling
  whiteScanner(1);        // Run white scanner effect for 1 round
  whitePulse(2);          // Run white pulse effect for 2 rounds
  blueFadeIn();           // Blue fade-in effect
  setAllBlue();           // Set all LEDs to blue permanently
}

void loop() {
  // Empty loop, no continuous effect needed
}

// Function for spark effect
void sparkEffect(int duration, int chanceOfSparkle) {
  int startTime = millis();
  while (millis() - startTime < duration) {
    strip.clear(); // Turn off all LEDs before updating
    
    // Randomly decide which LEDs to sparkle
    for (int i = 0; i < NUM_LEDS; i++) {
      if (random(100) < chanceOfSparkle) {
        strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set random LED to full white for sparkle
      }
    }

    strip.show();
    delay(50); // Delay between updates
  }
  strip.clear(); // Ensure all LEDs are off after the effect
  strip.show();
}

// Function for the white scanner effect
void whiteScanner(int rounds) {
  for (int r = 0; r < rounds; r++) {
    // Forward scan (LEDs won't stay active)
    for (int i = 0; i < NUM_LEDS; i++) {
      strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
      strip.show();
      delay(50);
      if (r == 0) {
        strip.setPixelColor(i, strip.Color(0, 0, 0)); // Turn off current LED after lighting up
      }
    }

    // Backward scan
    for (int i = NUM_LEDS - 1; i >= 0; i--) {
      strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set current LED to full white
      strip.show();
      delay(50);
      if (r == 0) {
        strip.setPixelColor(i, strip.Color(0, 0, 0)); // Turn off current LED after lighting up
      }
    }

    // Ensure all LEDs are fully lit at the end of the scan
    for (int i = 0; i < NUM_LEDS; i++) {
      strip.setPixelColor(i, strip.Color(255, 255, 255)); // Set all LEDs to full white
    }
    strip.show();
    delay(500); // Brief pause before starting the turn-off sequence

    // Turn off LEDs in steps of four
    for (int i = 0; i < NUM_LEDS; i += 4) {
      for (int j = i; j < i + 4 && j < NUM_LEDS; j++) {
        strip.setPixelColor(j, strip.Color(0, 0, 0)); // Turn off current LED
      }
      strip.show();
      delay(200); // Delay between turning off each group of four LEDs

      // Optional: Print the current state for debugging  
      Serial.print("Turning off LEDs from ");
      Serial.print(i);
      Serial.print(" to ");
      Serial.println(i + 4); 
    }
  }
}

// Function for white pulse effect
void whitePulse(int rounds) {
  for (int r = 0; r < rounds; r++) {
    // Brighten up
    for (int brightness = 0; brightness <= 255; brightness += 5) {
      for (int i = 0; i < NUM_LEDS; i++) {
        strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs brighten up
      }
      strip.show();
      delay(10); // Speed of brightening up
    }
    
    // Dim down
    for (int brightness = 255; brightness >= 0; brightness -= 5) {
      for (int i = 0; i < NUM_LEDS; i++) {
        strip.setPixelColor(i, strip.Color(brightness, brightness, brightness)); // All LEDs dim down
      }
      strip.show();
      delay(10); // Speed of dimming down
    }
  }
}

// Function for blue fade-in effect
void blueFadeIn() {
  for (int brightness = 0; brightness <= 255; brightness++) {
    for (int i = 0; i < NUM_LEDS; i++) {
      strip.setPixelColor(i, strip.Color(0, 0, brightness)); // Fade in blue
    }
    strip.show();
    delay(40); // Smooth fade-in effect
  }
}

// Function to set all LEDs to blue permanently
void setAllBlue() {
  for (int i = 0; i < NUM_LEDS; i++) {
    strip.setPixelColor(i, strip.Color(0, 0, 255)); // Set LED to full blue
  }
  strip.show();
}

**Comment by bujo-hub** [comment]
1
